crates/wasmtime/src/runtime/vm/instance/allocator/pooling.rs

```diff
@@ -652,23 +664,30 @@ unsafe impl InstanceAllocator for PoolingInstanceAllocator {
         ty: &wasmtime_environ::Memory,
         memory_index: Option<DefinedMemoryIndex>,
     ) -> Result<(MemoryAllocationIndex, Memory)> {
-        // FIXME(rust-lang/rust#145127) this should ideally use a version of
-        // `with_flush_and_retry` but adapted for async closures instead of only
-        // sync closures. Right now that won't compile though so this is the
-        // manually expanded version of the method.
-        let e = match self.memories.allocate(request, ty, memory_index).await {
-            Ok(result) => return Ok(result),
-            Err(e) => e,
-        };
+        self.live_memories.fetch_add(1, Ordering::AcqRel);
+        async {
+            // FIXME(rust-lang/rust#145127) this should ideally use a version of
+            // `with_flush_and_retry` but adapted for async closures instead of only
+            // sync closures. Right now that won't compile though so this is the
+            // manually expanded version of the method.
+            let e = match self.memories.allocate(request, ty, memory_index).await {
+                Ok(result) => return Ok(result),
+                Err(e) => e,
+            };
 
-        if e.is::<PoolConcurrencyLimitError>() {
-            let queue = self.decommit_queue.lock().unwrap();
-            if self.flush_decommit_queue(queue) {
-                return self.memories.allocate(request, ty, memory_index).await;
+            if e.is::<PoolConcurrencyLimitError>() {
+                let queue = self.decommit_queue.lock().unwrap();
+                if self.flush_decommit_queue(queue) {
+                    return self.memories.allocate(request, ty, memory_index).await;
+                }
             }
-        }
 
-        Err(e)
+            Err(e)
+        }
+        .await
+        .inspect_err(|_| {
+            self.live_memories.fetch_sub(1, Ordering::AcqRel);
+        })
```

Comment:
You'll want to switch this to using a `Drop` implementation because otherwise this won't be correctly accounted for if the future is cancelled/dropped.

Reply:
Ah good point; I didn't think about this hard enough when I rebased on #11470. I'll just switch to incrementing on Ok anyway.